routing/missioncontrol.go

```diff
@@ -279,9 +282,166 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
 
 // ResetHistory resets the history of missionControl returning it to a state as
 // if no payment attempts have been made.
-func (m *missionControl) ResetHistory() {
+func (m *MissionControl) ResetHistory() {
+	m.Lock()
+	defer m.Unlock()
+
+	m.history = make(map[route.Vertex]*nodeHistory)
+
+	log.Debugf("Mission control history cleared")
+}
+
+// getEdgeProbability is expected to return the success probability of a payment
+// from fromNode along edge.
+func (m *MissionControl) getEdgeProbability(fromNode route.Vertex,
+	edge EdgeLocator, amt lnwire.MilliSatoshi) float64 {
+
 	m.Lock()
-	m.failedEdges = make(map[EdgeLocator]time.Time)
-	m.failedVertexes = make(map[route.Vertex]time.Time)
-	m.Unlock()
+	defer m.Unlock()
+
+	// Get the history for this node. If there is no history available,
+	// assume that it's success probability is 1. After the attempt new
+	// information becomes available to adjust this probability.
+	nodeHistory, ok := m.history[fromNode]
+	if !ok {
+		return 1
+	}
+
+	return m.getEdgeProbabilityForNode(nodeHistory, edge.ChannelID, amt)
+}
+
+// getEdgeProbabilityForNode estimates the probability of successfully
+// traversing a channel based on the node history.
+func (m *MissionControl) getEdgeProbabilityForNode(nodeHistory *nodeHistory,
+	channelID uint64, amt lnwire.MilliSatoshi) float64 {
+
+	// Calculate the last failure of the given edge. A node failure is
+	// considered a failure that would have affected every edge. Therefore
+	// we insert a node level failure into the history of every channel.
+	var lastFailure time.Time
+	if nodeHistory.lastFail != nil {
+		lastFailure = *nodeHistory.lastFail
+	}
+
+	// Take into account a minimum penalize amount. For balance errors, a
+	// failure may be reported with such a minimum to prevent too aggresive
+	// penalization.
+	channelHistory, ok := nodeHistory.channelLastFail[channelID]
+	if ok && channelHistory.minPenalizeAmt <= amt {
+		if channelHistory.lastFail.After(lastFailure) {
+			lastFailure = channelHistory.lastFail
+		}
+	}
+
+	timeSinceLastFailure := m.now().Sub(lastFailure)
+
+	// Calculate success probability. It is an exponential curve that brings
+	// the probability down to zero when a failure occurs. From there it
+	// recovers asymptotically back to 1. The rate at which this happens is
+	// controlled by the penaltyHalfLife parameter.
+	exp := -timeSinceLastFailure.Hours() / m.cfg.PenaltyHalfLife.Hours()
+	probability := 1 - math.Pow(2, exp)
```

Comment:
May be useful to log the probability on the trace level for debugging purposes (with some of the surrounding context).

Reply:
That will be a lot of logging. One line per evaluated edge.

Reply:
I think that's okay on trace

Reply:
Log added